File: haskell/pig-latin/src/PigLatin.hs
{-# LANGUAGE LambdaCase #-}

module PigLatin
  ( translate,
  )
where

translate :: String -> String
translate = unwords . map translateWord . words

translateWord :: String -> String
translateWord word@('x' : 'r' : _) = word <> "ay"
translateWord word@('y' : 't' : _) = word <> "ay"
translateWord word =
  case break (`elem` vowels) word of
    ("q", 'u' : rest) -> rest <> "quay"
    (consonants@(_ : "q"), 'u' : rest) -> rest <> consonants <> "uay"
    ("", 'y' : rest) -> rest <> "yay"
    (consonants, rest@('y' : _)) -> rest <> consonants <> "ay"
    (consonants, rest) -> rest <> consonants <> "ay"

vowels :: [Char]
vowels = "aeiouy"

File: julia/pig-latin/pig-latin.jl
"""
    translate(phrase::AbstractString)::String

Translate a `phrase` from English to Pig Latin.
"""
function translate(phrase::AbstractString)::String
    join(map(translate_word, split(phrase, " ")), " ")
end


"""
    translate_word(word::AbstractString)::String

Translate a `word` from English to Pig Latin.
"""
function translate_word(word::AbstractString)::String
    if occursin(r"^(?:xr|yt|[aeiou]+)\w+$", word)
        return word * "ay"
    end

    m = match(r"^([^aeiouy]*q)u(\w+)$", word)
    if m ≠ nothing
        return join(reverse(m.captures), "") * "uay"
    end

    m = match(r"^y(\w+)$", word)
    if m ≠ nothing
        return m.captures[1] * "yay"
    end

    m = match(r"^([^aeiouy]+)(y?\w+)$", word)
    if m ≠ nothing
        return join(reverse(m.captures), "") * "ay"
    end

    throw(DomainError(word, "word of unknown pattern"))
end
